typing/jkind.ml

```diff
@@ -1869,26 +1965,20 @@ let get_layout jk : Layout.Const.t option = Layout.get_const jk.jkind.layout
 
 let extract_layout jk = jk.jkind.layout
 
-let upper_bound_for_axis (type a) ~type_equal ~jkind_of_type ~(axis : a Axis.t)
-    jk =
-  With_bounds.extend_bound ~axis ~type_equal ~jkind_of_type
-    ~bound:(Mod_bounds.get jk.jkind.mod_bounds ~axis)
-    jk.jkind.with_bounds
-
-let get_modal_upper_bounds ~type_equal ~jkind_of_type jk : Alloc.Const.t =
-  let upper_bound_for_axis (type a) (axis : a Axis.t) =
-    upper_bound_for_axis ~type_equal ~jkind_of_type ~axis jk
-  in
-  { areality = upper_bound_for_axis (Modal (Comonadic Areality));
-    linearity = upper_bound_for_axis (Modal (Comonadic Linearity));
-    uniqueness = upper_bound_for_axis (Modal (Monadic Uniqueness));
-    portability = upper_bound_for_axis (Modal (Comonadic Portability));
-    contention = upper_bound_for_axis (Modal (Monadic Contention))
+let get_modal_upper_bounds (type l r) ~jkind_of_type (jk : (l * r) jkind) :
+    Alloc.Const.t =
+  let jk = normalize ~require_best:false ~jkind_of_type jk in
```

Comment:
I think soundness here requires that the result of `normalize` has no with-bounds. Should we assert that? Or maybe `normalize ~require_best:false` returns an lr-kind?

Reply:
In the next PR (https://github.com/ocaml-flambda/flambda-backend/pull/3525), @liam923 refactors `normalize` to return an lr-kind in the not_best case. Plausibly it's still worth adding the assert, but I think it should happen there if so?

Reply:
In my PR, it using the `normalize_mode` gadt to guarantee this via types, and then it matches the result's with-bounds against `No_with_bounds` to assert this in the type system. But now that `With_bounds`'s map can be empty, this type assertion isn't as powerful as it used to be.

Reply:
> now that `With_bounds`'s map can be empty

it can representationally... but it should never actually be empty